rfc/rfc-0014-binding-injection.md

```diff
@@ -0,0 +1,131 @@
+
+A duck type for exposing a binding is defined in RFC 0002. While not as wide spread, the Stream resource currently implements this duck type. Combining the Binding and PodSpecable duck types allows for a generic means to inject binding metadata and secrets into a consuming workload. A new CRD can express this injection:
+
+```yaml
+apiVersion: bindings.projectriff.io/v1alpha1
+kind: Injection
+metadata:
+  name: my-binding-injection
+spec:
+  binding:
+    apiVersion: streaming.projectriff.io/v1alpha1
+    kind: Stream
+    name: my-stream
+  containers:
+  - user-container
+  subject:
+    apiVersion: knative.projectriff.io/v1alpha1
+    kind: Deployer
+    name: my-deployer
+status:
+  observedGeneration: 1
+  conditions: []
+```
+
+- `.spec.binding` is an ObjectReference to a resource implementing the binding duck type
+- `.spec.containers` optionally restricts which containers are augmented, defaults to all containers
+- `.spec.subject` is an ObjectReference to a resource implementing the podspecable duck type
+- the subject and target much be in the same namespace as the injection resource
+
+The subject resource will be injected with the binding metadata and secret.
+
+```yaml
+apiVersion: streaming.projectriff.io/v1alpha1
+kind: Stream
+metadata:
+  name: my-stream
+spec: {}
+status:
+  binding:
+    metadataRef:
+      name: my-stream-binding-metadata
+    secretRef:
+      name: my-stream-binding-secret
+```
+
+```yaml
+apiVersion: knative.projectriff.io/v1alpha1
+kind: Deployer
+metadata:
+  name: before
+spec:
+  template:
+    spec:
+      containers:
+      - name: user-container
+        image: square
+```
+
+```yaml
+apiVersion: knative.projectriff.io/v1alpha1
+kind: Deployer
+metadata:
+  name: after
+spec:
+  template:
+    spec:
+      containers:
+      - name: user-container
+        image: square
+        env:
+        - name: CNB_BINDINGS
+          value: /var/riff/bindings
+        volumeMounts:
+        - mountPath: /var/riff/bindings/my-binding-injection/metadata
+          name: my-binding-injection-metadata
+          readOnly: true
+        - mountPath: /var/riff/bindings/my-binding-injection/secret
+          name: my-binding-injection-secret
+          readOnly: true
+    volumes:
+    - name: my-binding-injection-metadata
+      configMap:
+        name: my-stream-binding-metadata
+    - name: my-binding-injection-secret
+      secret:
+        secretName: my-stream-binding-secret
+```
+
+Notes:
+- if the `CNB_BINDINGS` env var is already defined, it must be respected
+  - otherwise it must be defined
+- volume names must not collide with an existing volume
+- volume mount paths must be under the CNB_BINDINGS value
+- the mount directory within CNB_BINDINGS must be unique
```

Comment:
I think it should have the same name as the binding object. It looks like `my-binding-injection` in this example was pulled out of a hat. Or am I missing something?

Reply:
I, OTOH think it should have the name of the *Injection* object, or something derived from it.
A single binding can be injected multiple times with different "roles". For example, a single Stream can be used as different inputs (or different outputs) of a Processor.

Have a look at the http-source [PR](https://github.com/projectriff/http-source/pull/9#issue-347115960) for cardinalities that make sense IMO.

Reply:
I think that's what I meant. A `Binding` is "global" (could be shared by multiple `Injections`, or even multiple apps). I don't have the same hang up about using the name `binding` to refer to the "injection" though.

Reply:
Since the path for the volume mount must be unique, there will necessarily need to be a bit of massaging at runtime to avoid collisions. Don't get hung up on what the name is.